### PR TITLE
fix Dynamic properties do not work

### LINKED ISSUE #3391: [Adaptive[ Dynamic properties do not work
dialog[turn.abc].score where turn.abc resolves to string 'result' does not work.

### PR DESCRIPTION
**Pull request by Danieladu**
close #3391 

When we evaluate such expression like "dialog[a].b" with the scope implement by `DialogStateManager`,
The steps to get the result:
1. call DialogStateManager.TryGetValue
2.TransformPath path to get the right format path
3.ResolveMemoryScope to get the memory scope and remaining path
...

The issue occurs in step3, when `dialog['a'].b` comes in, `ResolveMemoryScope`  would separate the path into two parts, first part is `dialog['a']`, and the second part is `b`, but `dialog['a']` is not a valid scope, `dialog` is. 

How to reproduce:
build the below dialog structure:
```
"actions": 
[
                    {
                      "$kind": "Microsoft.SetProperty",
                      "property": "user.name.firstname",
                      "value": "dialog.jack"
                    },
                    {
                      "$kind": "Microsoft.SetProperty",
                      "property": "turn.nameproperty",
                      "value": "name"
                    },
                    {
                      "$kind": "Microsoft.SetProperty",
                      "property": "=user[turn.nameproperty].firstname",
                      "value": "daniel"
                    },
                    {
                      "$kind": "Microsoft.SendActivity",
                      "activity": "new firstname is ${dialog.jack}"
                    }
  ]
```

expect set value "daniel" into property "dialog.jack", but failed.

 Expectation:
Separate path with both open square bracket and dot to get the right memoryscope.

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Memory/DialogStateManager.cs [hunk 1]
```diff
@@ -123,14 +123,30 @@ string IMemory.Version()
         public virtual MemoryScope ResolveMemoryScope(string path, out string remainingPath)
         {
             var scope = path;
+            var sepIndex = -1;
             var dot = path.IndexOf(".");
-            if (dot > 0)
+            var openSquareBracket = path.IndexOf("[");
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Memory/DialogStateManager.cs]: >var openSquareBracket = path.IndexOf("[") [](start = 12, length = 41)

No description of what this change is about.
no unit tests testing the new functionality
Please add unit tests and comments.

Comment (Danieladu) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Memory/DialogStateManager.cs]: yes, I have added the detailed descriptions for this pr. 
But, I do now know where to put the unit test of DialogStateManager,  it doesn't seem to have unit tests originally.

**Comment by vishwacsena** [comment]
Tagging R8.

**Comment by boydc2014** [comment]
> Verified functionally that this fixes the regression w/ dynamic property access.

It's not exactly a regression, it's a long live bug i believe. The issue is that, DSM do path resolving with two step
* step 1 : identify the scope, for example, dialog.abc, will identify as dialog scope, then path will becomes "abc" 
* step 2 : resolve value inside scope using the path left

The issue is in step 1 that we are only searching first "." to identify scope, so for "dialog["abc"]", it wouldn't correctly identify as dialog scope. The fix is to search "." and "[" both. @tomlm will this help explain the issue?

**Pull request merged**